flecs_ecs_tracing/src/perf_trace.rs

```diff
@@ -0,0 +1,130 @@
+use crate::{metadata, util::*};
+use core::ffi;
+use std::{
+    borrow::Cow,
+    collections::HashMap,
+    sync::{LazyLock, OnceLock, RwLock},
+};
+use tracing_core::{span, Dispatch, Level, Metadata};
+
+/// Log level for performance tracing spans
+pub(crate) static SPAN_LEVEL: OnceLock<Level> = OnceLock::new();
+
+/// Registry of dynamic tracing spans generated from Flecs
+static REGISTRY: LazyLock<RwLock<Registry>> = LazyLock::new(Default::default);
+
+/// Registry of dynamic tracing spans generated from Flecs
+struct Registry {
+    spans: HashMap<
+        String,
+        (tracing_core::span::Id, &'static Metadata<'static>),
+        fxhash::FxBuildHasher,
+    >,
+}
+
+impl Default for Registry {
+    fn default() -> Self {
+        Self {
+            spans: HashMap::with_capacity_and_hasher(256, Default::default()),
+        }
+    }
+}
+
+pub(crate) fn init() {
+    LazyLock::force(&REGISTRY);
+}
+
+fn ensure_span(
+    dispatch: &Dispatch,
+    filename: Option<Cow<'_, str>>,
+    line: usize,
+    name: Cow<'_, str>,
+) -> (span::Id, &'static Metadata<'static>, bool) {
+    let mut registry = REGISTRY.write().unwrap();
+
+    // Double-checked locking (maybe another thread added this span between read unlock and write lock)
+    if let Some((existing, metadata)) = registry.spans.get(name.as_ref()) {
+        return (existing.clone(), metadata, false);
+    }
+
+    // "C++-style" names allow Tracy UI to auto-collapse names in short spans
+    // This conversion is done late so that the fast path can avoid allocating;
+    // the unmodified name is used as the hashmap key in the registry.
+    let meta_name = name.replace(".", "::");
+
+    // Don't bother caching metadata for these; there's (currently) a 1-to-1 correspondence between spans and metadata.
+    let (metadata, _) = metadata::MetadataRequest {
+        kind: metadata::MetadataKind::Span,
+        name: Cow::Owned(meta_name),
+        filename,
+        line: line.try_into().ok(),
+        level: *SPAN_LEVEL.get().unwrap_or(&Level::INFO),
+    }
+    .register(dispatch);
+
+    // Flecs doesn't send any parameters (yet)
+    let values = metadata.fields().value_set(&[]);
+    let attributes = span::Attributes::new(metadata, &values);
+
+    // Tell `tracing` subscribers about the new span
+    let span = dispatch.new_span(&attributes);
+
+    registry
+        .spans
+        .insert(name.into_owned(), (span.clone(), metadata));
+
+    (span, metadata, true)
+}
+
+fn get_span(
+    dispatch: &Dispatch,
+    filename: Option<Cow<'_, str>>,
+    line: usize,
+    name: Cow<'_, str>,
+) -> (span::Id, &'static Metadata<'static>, bool) {
+    if let Some((existing, metadata)) = REGISTRY.read().unwrap().spans.get(name.as_ref()) {
```

Comment:
This and another instance of a similar code pattern feel like they should be chained functions rather than an if / early return. But maybe not?